<div align="center"><h1><b>Implementación del Método Numérico de Simpson</b></h1></div>
<div align="center"><h1><b>Para el cálculo de Áreas Curvas</b></h1></div>
<div align="center"><h1><b>Utilizando el Lenguaje C y Paradigmas de Paralelización</b></h1></div>
<div align="center"><h3>Nicolás Cuello Suárez - Curso de Programación de Alto Desempeño</h3></div>
<div align="center"><h3>Facultad de Ciencias Básicas - Universidad de Medellín</h3></div>


---

## **El Paradigma del Cálculo Integral.**


En el contexto de la formación para la **ingeniería** y la **ciencia**, una de las habilidades más importantes que se deben desarrollar es la capacidad de **abstracción** y entendimiento del mundo para la resolución de problemas; específicamente para la observación de fenómenos que ocurren en la realidad de forma constante. Ese es, precisamente, uno de los objetivos de la ciencia. A lo largo de este proceso de formación, nos podemos encontrar con diversos campos del conocimiento que nos permiten acercarnos a ese fin: **El álgebra y sus vertientes**, **La computación**, **La Biología**, etc. Sin embargo, si ponemos la mira en el área del **cálculo**, podemos encontrar las bases para el modelamiento del mundo como lo conocemos.

En el uso de las **mátematicas**, del **análisis funcional** y de las herramientas que el **cálculo** nos brinda, podemos modelar procesos complejos que hay en nuestro universo, como por ejemplo, la dinámica de *cuerpos planetarios*, el comportamiento de *fluidos* y el funcionamiento de los pequeños *procedimientos químicos* que hace la materia que compone nuestra existencia.

Hemos logrado hacer avances en muchos campos, gracias a las bases de abstracción que el **cálculo**, como objeto de estudio, brinda. Dentro de esta área, podemos notar una base fundamental para el análisis de funciones, y es el hallazgo de áreas a partir del proceso de **integración**, una operación fundamental que nace desde las preguntas:

- **¿Cómo hallar la magnitud de regiones que NO son medibles por los procesos geométricos simples?**
- **¿Cómo se pueden hallar áreas de regiones euclidianas curvas?**

A partir de los conceptos básicos de **integración**, logramos responder a este tipo de preguntas, pues, la **integral definida** es, en escencia, hallar el área contenida por una o más funciones en un intervalo específico. Ahora bien:
- **¿De qué forma se hace este proceso en "papel"?** 
- Centrándonos en el ámbito computacional, **¿Hay alguna manera de programar algoritmos computables que aproximen esos cálculos?** 
- Y algo importante importante (teniendo en cuenta nuestro enfoque de **Programación de Alto Desempeño**):
> 1. **¿De qué forma se pueden optimizar esos cálculos utilizando procesos como la paralelización en diferentes hilos, en el lenguaje C con OpenMP?**
> 2. **¿Qué diferencias hay entre la integración serial y paralelizada?**

Todas estas preguntas serán respondidas en este Notebook, explicando a profundidad cada uno de los procesos **matemáticos** necesarios para implementar estrategias de integración, cada algoritmo implementado y cada estrategia de paralelización pensada, con la necesidad de comparar resultados y eficiencia entre versiones ejecutadas con directivas de paralelización y con versiones que no las tienen.

---

## **Las Bases del Cálculo de Áreas.**

Si estamos en un contexto universitario, el "ABC" del ingeniero corresponde a el conocimiento en los inicios paradigmáticos del cálculo, es decir, todo se construye a partir de principios tan claves como lo son los procesos de **derivación** y **antiderivación**. Formalmente, conocemos a la derivada como el cambio infinitamente inmediato que existe en una función univariable, vemos que la definición formal de la derivada corresponde a:

$$f'(x) = \lim_{h \to 0} \frac{f(x + h) - f(x)}{h}$$


Es decir, un límite de $h$ a $0$ en el que $h$ corresponde al cambio **(en $x$)** de la función; este proceso nos permite hallar la **recta tangente** a un punto específico en una función, pues la diferencia se hace infinitamente pequeña. A partir de esto, nacen múltiples métodos de derivación para hallar las derivadas de cualquier función (desde funciónes *polinómicas* a *trigonométricas*, *inversas trigonométricas*, etc.); Esto nos hace notar algo, y es que, al derivar una función, estas aplicando ciertas reglas que idealmente, pueden deshacerse, es decir, al derivar, **estás transformando una función a partir de reglas**. 

Si hacemos un ejercicio mental breve, podemos concluir que, si podemos derivar una función, podemos ***antiderivarla*** luego. Así pues, tenemos otra herramienta clave, la **antiderivada (o primitiva de una función)** como operación inversa a la **derivada**, que se define como el proceso por el que se halla una función primitiva **$F(x)$**, que al derivarla, logramos nuevamente la función original **$f(x)$** tal que:

$$F'(x) = f(x)$$

Y si **$F(x)$** es una **antiderivada** particular, podemos expresar la operación de **antiderivada** mediante la **integral indefinida** que representa una **familia de curvas**:

$$\int f(x) \ dx = F(x) + C$$

Así pues, construimos herramientas poderosas para el **análisis**, sin embargo, todo esto está basado en el uso del álgebra para la transformación de funciones definidas; no logramos una pista para lograr valores numéricos que representen áreas en intervalos específicos, hasta ahora, es mero uso del álgebra simbólica para operar funciones tanto **derivándolas** como **antiderivándolas**.

Para responder a esta necesidad, nace una herramienta mucho más poderosa aún, la **integral definida**, que nos permite, básicamente, hallar el **área** que existe debajo de la **curva o recta** de una función **$f(x)$**, en un **intervalo específico**. 

**¿Cómo se calcula esta área?**: muy simple, a partir de las herramientas construidas anteriormente: El método clásico y de perogrullo que habitualmente es enseñado, es el de, hallar la **integral indefinida** de la función cuya área se quiere encontrar, y posterior a eso, **evaluar** en unos límites de integración definidos (**$a$** y **$b$**) que corresponden a los extremos del intervalo a analizar, tal que:



$$\int_{b}^{a} f(x) \ dx$$

Esto funciona porque es el equivalente a una herramienta matemática que estabamos pasando por alto y es crucial para el cálculo de áreas: **La Sumatoria de Riemann**, una **sumatoria básica** en la cual dividimos el intervalo a calcular en un número de **rectángulos** específico, y sumamos el área (base X altura) de cada uno de ellos, siendo la **altura**, el valor de la función en ese punto y la base, la base del rectángulo creado; entre más divisiones (rectángulos) halla, vamos a lograr un resultado más preciso y apegado a la realidad. Así pues, pasamos de cálculos más complicados algebráicamente (como los procesos de **derivación** y **antiderivación**), a algo mucho más simple, como lo es la sumatoria de áreas más fáciles de calcular. Podemos definir, entonces, a la **integral definida** como la **Suma de Riemann** cuando la base de esos **rectangulos** tiende a $0$, es decir, cuando la cantidad de rectángulos tiende a infinito:

$$\int_{b}^{a} f(x) \ dx = \lim_{n \to \infty} \sum_{i = 1}^n f(x_i^*) \Delta x$$

Esto es contrastado y funcional, un recurso matemático resolutivo y útil para multiples necesidades: 

- Hallar **áreas** contenidas entre una o diferentes curvas.
- Hallar **volúmenes** de sólidos de revolución.
- Hallar el **trabajo** necesario para elevar fluídos a cierta altura.
- Hallar la **fuerza hidrostática** que ejerce un fluído sobre x figura.

Así pues, definimos a la integral definida como una sumatoria de áreas infinitesimalmente pequeñas. Sin embargo, esto nos hace crear nuevas preguntas:

1. **¿Todas las áreas pueden hallarse de esa forma?**
2. **¿Todas las funciones tienen una antiderivada primitiva que puede expresarse?**
3. **¿Es esta la forma más eficiente de hallar áreas?**
4. **¿Se pueden computar estos métodos?, ¿Son computacionalmente óptimos?**

---

## **La Integración Numérica: Una Optimización Pura.**

Lo visto anteriormente es una generalización matemática acertada y descrita por los genios más grandes de la historia; es funcional para todos y cada uno de los casos, todas las áreas pueden hallarse utilizando lo anterior. **¿O no?**, probémoslo, utilicemos una función arbitraria cualquiera para hallar su antiderivada y evaluar en límites. Probemos con una función simple:

$$f(x) = \cos(x²)$$ $$\int \cos(x²) \ dx$$

**¿Podemos utilizar la integral indefinida?**, fácil, sustitución simple, **¿o no?**: no, pues al utilizar a $x²$ como $u$ y derivarla, la integral queda en dos variables, no podemos liberarnos. 

**¿Sustitución trigonométrica?**, no, pues no hay identidad que permita convertir a la integral en una forma que permita esta sustitución. 

**¿Integración por Partes?**, no, entramos en un bucle infinito.

Entonces, ¿cómo lo integramos?; ahí, acabamos de entrar a un factor importante, y es que existen funciones cuya **antiderivada existe**, pero no puede ser expresada con una fórmula finita conocida fácilmente. Es decir, sabemos que existe un área debajo de la curva de $\cos(x²)$, sin embargo, no existe una combinación básica de funciones cuya derivada sea $\cos(x²)$. Entramos al campo de las **Integrales NO elementales**.

Además, pensemos en algo, si queremos optimizar la resolución de un problema, lo más facil es programarlo, así podemos generalizar y conseguir respuestas de forma mucho más óptima. Entonces, ¿podemos computabilizar este problema?, según lo que hemos dicho, necesitaríamos poder programar antiderivadas. Eso existe, se llama **Álgebra Computacional**, sin embargo, preguntémonos, **¿es esta la forma más óptima de solucionar problemas de grandes volúmenes de datos?**, **¿qué tan fácil es programar álgebra simbólica?**.

Ambas respuestas nos dan un mal augurio, puesto que para programar **álgebra simbólica computacional**, se deben manipular caracteres, reglas lógicas y transformaciones que consumen tiempo de procesamiento y memoria; simplemente para antiderivar una función como:

$$f(x) = \sin²(x)\cos³(x)$$

Se requiere de múltiples transformaciones, identidades y sustituciones, lo que hace de estos algoritmos algo complejo e ineficiente de manipular, por todas las consideraciones que se deben tener. Por eso, herramientas como ***Mathematica*** o ***SymboLab*** son tan pesadas. Además, se deben implementar algoritmos sumamente complejos, extensos y poco ortodoxos. 

Es por eso, que existe una respuesta óptima, funcional, y muy buena ante estas dificultades: **el cálculo numérico**, un conjunto de algoritmos que utilizan operaciones aritméticas simples como la **suma** o **la multiplicación** para lograr resultados muy bien aproximados y convergentes para problemas matemáticos mucho más complejos. Se debe tener un análisis más simplista de los problemas para resolverlos, sin embargo, pueden ser computacionalmente mucho más óptimos y fáciles de resolver, pues se basan en operaciones básicas y más sencillas para un procesador.

Por ejemplo, la sumatoria de Riemann nos permitiría hallar el área debajo de la curva en un intervalo para **Integrales No Elementales**, pues es un problema fácil de computar, una sumatoria de las áreas de $n$ cantidad de rectángulos. Así pues, vemos que el problema anteriormente planteado se soluciona de una forma simple: **utilizando operaciones básicas para discretizar procesos complejos**, incluso, podemos implementar directivas de paralelización para una resolución de esa índole, sin embargo, debemos, nuevamente, preguntarnos lo siguiente:

- **¿Es esta resolución para el cálculo de áreas la más óptima en términos computacionales?...**

Hagamos un ejercicio para responder a la disyuntiva anterior, probemos calcular el área de $f(x) = e^{-x²}$ en el intervalo cerrado de $[-1, 1]$, tal que:

$$\int_{-1}^{1} e^{-x²}\ dx$$

Para ejecutar este cálculo, debemos primero pensar en el algoritmo que nos va a permitir realizarlo; en este caso, podemos plantearlo de una manera muy simple; si implementamos un código serial, utilizando C, para hallar el área debajo de la curva de la función dada en el intervalo solicitado. 

Lo hacemos de la siguiente manera:

In [15]:
//%ldflags: -lm
// Flag utilizada para incluir -libm
#include <stdio.h>
#include <math.h>
#include <omp.h>
#define real_value  

double function(double x) {
    return exp(-(x*x));
}

double calculate_integral_using_riemann_sum(double inferior_limit, double superior_limit, double rectangles) {
    double sum = 0;

    if (inferior_limit < superior_limit) {
        double distance_between_limits = superior_limit - inferior_limit;
        double split = distance_between_limits / rectangles;

        while (inferior_limit < superior_limit) {
            inferior_limit += split;
            sum += function(inferior_limit) * split;
        }

        return sum;
    } else if (inferior_limit == superior_limit) {
        return 0;
    } else {
        double distance_between_limits = inferior_limit - superior_limit;
        double split = distance_between_limits / rectangles;

        while (superior_limit < inferior_limit) {
            superior_limit += split;
            sum += function(superior_limit) * split;
        }

        return -sum;
    }
}

int main() {
    double rectangles[6] = {20.0, 60.0, 500.0, 1000.0, 100000.0, 1000000.0};
    double inferior_limit = -1.0;
    double superior_limit = 1.0;

    FILE *file = fopen("data.dat", "w");
    fprintf(file, "iteraciones tiempo_ms resultado error\n");

    int i = 0;
    for (i = 0; i < 6; i++) {
        double initial_time = omp_get_wtime();
        double area = calculate_integral_using_riemann_sum(inferior_limit, superior_limit, rectangles[i]);
        double error = fabs(real_value - area);
        double final_time = omp_get_wtime();
        double time_passed = (final_time - initial_time) * 1000.0;

        printf("Si hallamos el área de la función e^(-x²) en el intervalo [%.1f, %.1f] logramos: %.10f\n", inferior_limit, superior_limit, area);
        printf("Con %.1f iteraciones, el tiempo de ejecución fue de %f ms \n", rectangles[i], time_passed);
        printf("\n");

        fprintf(file, "%.0f %.6f %.10f %.10f\n", rectangles[i], time_passed, area, error);
    }

    fclose(file);
    return 0;
}

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.5222413202
Con 20.0 iteraciones, el tiempo de ejecución fue de 0.035567 ms 

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.5049710625
Con 60.0 iteraciones, el tiempo de ejecución fue de 0.004031 ms 

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.4936463036
Con 500.0 iteraciones, el tiempo de ejecución fue de 0.019140 ms 

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.4936477751
Con 1000.0 iteraciones, el tiempo de ejecución fue de 0.037508 ms 

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482656
Con 100000.0 iteraciones, el tiempo de ejecución fue de 3.919010 ms 

Si hallamos el área de la función e^(-x²) en el intervalo [-1.0, 1.0] logramos: 1.4936490014
Con 1000000.0 iteraciones, el tiempo de ejecución fue de 56.464780 ms 



Inicialmente, incluimos como cabeceras, el Estándar de entrada y salida de información `<stdio.h>` para poder visualizar la información realizada en el cálculo. Adicionalmente, añadimos la cabecera `<math.h>` para "importar" operaciones y expresiones matemáticas importantes para esta y otras pruebas. En este caso, lo utilizamos para poder utilizar la expresión `exp()`, que representa, básicamente, el número $e$ elevado a lo que haya en el paréntesis.

Definimos una función `double function(double x) {}` para definir la función o curva la cual será nuestro objeto de cálculo. Recibe un número $x$ y devuelve la imágen que da la función al evaluarla con ese número $x$; basicamente, con esta función hallamos la "altura" del rectángulo infinitesimal.

Aquí encontramos la parte más importante para el cálculo de la integración numérica. Es un algoritmo verdaderamente simple, implementamos una función:

`double calculate_integral_using_riemann_sum(double inferior_limit, double superior_limit, double rectangles)`

En la que, inicialmente, se reciben los límites superior e inferior de la función, además de la cantidad de divisiones que se desea hacer; se define una variable crucial: `double sum` cuya función es acumular el área de cada uno de los rectángulos.

Posterior a eso, se manejan 3 casos:

1. El **límite inferior** es efectivamente menor al **límite superior**: Inicialmente, se calcula la distancia entre los límites (restando el inferior al superior) y se guarda en una variable `double distance_between_limits`, se halla la medida de la base de cada uno de los rectángulos (dividiendo la medida hallada entre el total de rectángulos), se guarda en una variable `double split` y posteriormente, se entra en un ciclo `while` que se itera hasta que el límite inferior deja de ser menor que el superior (eso significaría que el intervalo ya fue pasado); en cada iteración se le va sumando la base del rectángulo a el límite inferior, se evalúa el punto (en x, `inferior_limit`) alcanzado en la función, se multiplica por la base del rectángulo y se acumula en la variable `sum`.

2. Los límites son **iguales**: Simplemente se retorna 0, pues el área es 0.

3. EL **límite inferior** es mayor al **límite superior**: El algoritmo es el mismo que en el caso contrario, solo que ahora la variable `superior_limit` debe alcanzar al límite inferior. El área se devuelve negativa por el propiedades de la **integral definida**.



Ahora, listo el algoritmo de integración numérica, podemos proceder con la ejecución del mismo; en la salida, vemos que a más iteraciones, el valor se va acercando cada vez más a $1.4936 \ [u²]$, que, en la práctica, es el valor de el área necesitada.

En la función `main() {}` podemos utilizar las funciones planteadas inicialmente, sin embargo, para efectos prácticos y de **análisis**, realizaremos el ejercicio con cantidades diferentes de rectángulos, para ver las diferencias de convergencia y tiempo de ejecución, a medida que se va aumentando el número de **iteraciones**; para realizar este análisis de forma visual, utilizaremos GNUPlot para crear visualizaciones concretas sobre las estadísticas de las iteraciones.

Inicializamos un arreglo con diferentes cantidades de rectángulos (desde **veinte** iteraciones hasta **un millón**), utilizamos variables para almacenar los límites de **integración**, luego realizamos un ciclo `for` en el que realizamos el cálculo para cada número de rectángulos en el arreglo `rectangles[5]`, midiendo el tiempo de ejecución con la función `omp_get_wtime()`. Veamos las gráficas creadas en base a los datos:



<div style="text_align = center;">
  <img src="tiempo_vs_iteraciones.png" width="400"/>
  <img src="convergencia_vs_iteraciones.png" width="400"/>
</div>

En la primera, comparamos el **número de divisiones del área**, vs el **tiempo de ejecución**; en síntesis, podemos ver algo que nos podíamos esperar, a iteraciones bajitas, el tiempo de ejecución es menor (a pesar de ese bajón de tiempo entre 20 y 60 rectángulos, pues son cantidades tan bajas que son despreciables), a medida de que hay más divisiones, hay un mayor volúmen operacional, por lo que la gráfica crece. Por cómo planteamos el algoritmo, podemos decir que tiene una complejidad algorítmica de $O(n)$, y esto lo podemos confirmar con los datos de salida, vemos una tendencia lineal en los datos; lo vemos y confirmamos, es un algoritmo computacionalmente eficiente y simple; la integración numérica es una operación no muy costosa si la planteamos de esta manera.

Sin embargo, debemos tener en cuenta otra cuestión importante, y es que, a medida de que las iteraciones son mayores **¿En qué punto empieza a ser convergente al valor que queremos lograr?**, **¿Cuándo deja de ser óptimo seguir iterando para exactitud?**. Esto podemos analizarlo con la segunda gráfica, pues comparamos la convergencia a el **valor buscado**, vs las **iteraciones**. Logramos ver lo que esperábamos inicialmente, a menos rectángulos, el área hallada es mucho menos exacta (con veinte iteraciones se llegó a un área de $1.522241 \ [u²]$, cuyo margen de error es grande), sin embargo, a medida que se aumentan los rectángulos, el área se acerca más y más a la deseada, luego de cien mil, se logró un resultado relativamente cercano a el resultado exacto de la integral.

Como vemos, logramos el algoritmo definitivo para la integración, pues es eficiente, rápido, llega a valores aproximados reales y computacionalmente es verdaderamente simple, pues su complejidad algorítmica es lineal ($O(n)$) y su complejidad espacial es aún menor; es, de forma clara, un algoritmo exacto, útil, sencillo y preciso... **¿o no?**.

Empecemos por algo muy importante, y es que, hay errores en la idea anterior. Riemann es un método de integración funcional, pero al evaluar su eficiencia, tiene fallos importantes:

1. El error depende **directamente** del paso, es decir al aproximar el área con escalones, el error se vuelve dependiente a la cantidad de divisiones que se hagan; es una idea simple, la premisa de la **Sumatoria de Riemann**, sin embargo, eso lo convierte en algo poco óptimo, pues al tener una dependencia tan directa al número de iteraciones, el algoritmo es, en realidad, poco eficiente, pues para obtener precisión debes someterte a un número de pasos contínuos verdaderamente grande.
2. Cada iteración **acumula** un error constante, pues depende directamente del paso, lo cual hace que incluso en iteraciones altas, halla cierto nivel de error en el cálculo.
3. Este método **converge excesivamente lento**, pues, como lo vimos, apenas llega a un resultado verdaderamente cercano luego de **cien mil** pasos, y a partir de él, la mejora de convergencia es verdaderamente lenta.

En síntesis, es un método que es **muy dependiente** a los pasos, y que necesita de **muchos**, pero **muchos** puntos para estabilizarse, a pesar de su simpleza. Además, dependiendo de como se sitúen los rectángulos debajo de la gráfica, se va a aproximar mejor o peor la curvatura, pues no es lo mismo que los rectángulos toquen a la gráfica con su vértice *superior derecho*, a que lo hagan con el *superior izquierdo*, o a que lo hagan con la *mitad de su base en el lado superior*, en todos los casos hay una diferencia radical en el cálculo.

Ahora, teniendo en cuenta todos los problemas que puede tener la **Sumatoria de Riemann** computacionalmente hablando, es válido preguntarse lo siguiente:

- **¿Hay alguna forma de solucionarlos?**
- **¿Es el único método de Integración Numérica que existe?**

---

## **Una Alternativa (Casi) Milagrosa: Simpson 1/3**

Para resolver las preguntas anteriores, hagamos un trabajo mental. Anteriormente, nos enfocábamos en dividir la función en partes pequeñas (con rectángulos), aproximar $f(x)$ por constantes en cada intervalo, y sumar cada una de las partes halladas. Preguntémonos, ¿esto funciona para aproximar la curvatura de la función en cierto punto?
 
Respuesta rápida, **no lo hace**, pues si el rectángulo toca en un punto a la curva, básicamente no se está teniendo en cuenta el comportamiento, concavidad y tendencia de la función en ese punto. Pensemos en algo simple, **¿Al derivar, qué se le está haciendo a una función?**. Veámoslo, dada una función:

$$f(x) = x⁴ + 2x³ + 9x² + 6$$

Si la derivamos utilizando la regla exponencial de la derivación para polinomios algebráicos de $n$ grado, nos queda:

$$f'(x) = 4x³ + 6x² + 18x$$

Ahora bien, **¿qué hicimos?**, hallamos la función que representa el cambio instantáneo de cualquier punto de $f(x)$, ¿no?; en escencia, sí, pero notemos algo más allá, el grado de $f(x)$ es cuatro, y el grado de $f'(x)$ es tres, básicamente, lo que hicimos fue **reducirle el orden** a una función. Homólogamente, si al **derivar** le restamos un grado a la función, si la **antiderivada** corresponde a el proceso contrario a la derivada (por el teorema fundamental del cálculo), corresponde, precisamente a **añadirle un grado** a una función, y una función, a mayor grado, mayor cantidad de puntos se necesitan para dibujarla. 

Por ejemplo, una función lineal **(Por ejemplo: $f(x) = 3x + 2$)** corresponde a una **línea recta**: solo necesitas dos puntos para gráficarla, una **función cuadrática** corresponde a una **parábola**: solo se necesitan 3 puntos para realizar su bosquejo, para una **función cúbica**, vemos una gráfica con forma de $S$, y se necesitan 4 puntos para dibujarla. ¿Se nota el patrón?, al aumentar la curvatura de una función, necesitamos un número de puntos equivalentes a el grado más uno para graficarla.

Ahora, **¿qué tiene que ver esto con lo mencionado anteriormente?**, sigamos en el ejercicio y pensémos en algo simple: **¿es mejor aproximar áreas curvas con rectas y figuras rectas simples, o con curvas?**, y más allá, **¿podemos utilizar curvas más simples, para hallar el área de curvas más complejas?**.

En esto pensó el matemático inglés **Thomas Simpson** años antes de 1743, cuando publicó formalmente un artículo en el que explicó el funcionamiento de un método de **Interpolación Cuadrática** para hallar áreas debajo de curvas. Este método de **integración numérica** lleva por nombre el **Método de Simpson 1/3**, y se caracteriza por utilizar una estrategia de discretización que consiste en el uso áreas más faciles de calcular **(específicamente, parábolas)** para llegar a el cálculo de áreas más difíciles, como lo es el área debajo de una curva compuesta; de esta manera, eliminamos el problema que anteriormente teníamos al utilizar **Riemann**, pues ya no se aproxima sin tener en cuenta la curvatura de la función en el intervalo, ya no se toman áreas definidas y rectas, primero se realiza un mejor proceso de **aproximación** de curvatura en la función antes de **integrar**, y eso cambia mucho el proceso numérico.

Para definir formalmente el método de **Simpson 1/3**, entendamos analítica y más concretamente cómo funciona. Utilicemos la integral anterior:

$$\int_{-1}^{1} e^{-x²} \ dx$$

Bajo este **paradigma**, como anteriormente se mencionó, lo que debemos hacer, en primera instancia, es dividir el intervalo (en este caso $[-1, 1]$) en un **número par de sub-intervalos** de **longitud simétrica** (es decir, dividir el intervalo en un número par de partes iguales), así, para cada par de sub-intervalos, se tiene un punto de inicio, un punto medio, y un punto final; entonces, utilizamos esos puntos para interpolar una **parábola** que acerque mucho su figura a la curvatura de la función en esa **sub-región**, hallamos el área debajo de la parábola con límites en el punto de inicio y final de ese par de intervalos. Luego, se aplica eso mismo para los siguientes pares de **sub-intervalos**, utilizando los puntos compartidos de cada par adyacente. 

Al final, se suman todas las **integrales** locales de **parábolas**, logrando una suma ponderada de los valores de la función en los puntos del mallado, reutilizando los puntos finales, como iniciales en el siguiente par. 

La clave está en reemplazar la función por una función por **tramos cuadráticos**. Así logramos una mayor eficiencia en la convergencia, pues no solo se aproxima el área, se aproxima la forma de la curva antes de integrar. Notar que, ejecutado de esta forma estamos hablando del método de **Simpson 1/3** de aplicación múltiple, puesto que en la aplicación simple de este método de integración, simplemente dividimos toda el área en dos intervalos, y utilizamos **una sola parábola** para integrar toda la región y hallar el área. Formalmente, podemos describir el método de la siguiente manera:

$$\int_a^b f(x) \ dx \approx \frac{h}{3} [f(x_0) + 4 \sum_{impares} f(x_i) \ + 2 \sum_{pares} f(x_i) + f(x_n)]$$

Y la entendemos de una forma en la que tenemos un intervalo $[a, b]$ y $n$ es el número (par) de **sub-intervalos** de **igual longitud** en los que se divide el intervalo original. Utilizamos $h$ para representar la longitud de cada una de las divisiones, y podemos escribirla como $\frac {b - a}{n}$, es decir, la longitud del intervalo dividida por el número par total de sub-intervalos definidos. 

Luego, $x_i$ corresponde a los nodos del intervalo, donde inicia como $x_i = a$, luego $x_i = a + h$, luego $x_i = a + 2$, y así sucesivamente hasta que $x_i = b$, evaluamos entonces la función en cada uno de esos nodos. Finalmente, debemos tener algo en cuenta, y es que **Simpson 1/3** no suma todas las partes de forma uniforme, pues no todas cumplen con el mismo rol geométrico:

- $f(x_0)$ y $f(x_n)$ corresponden a los **extremos**, y su peso es uno, pues solo participan en una **parábola** cada uno (la del inicio, y la del final).
- Luego, en los sub-intervalos interiores podemos ver puntos **impares** que tienen un mayor peso (cuatro), pues inciden directamente en el planteamiento de la parábola (son los **vértices** de la parábola auxiliar creada), y puntos **pares**, cuyo peso es menor (dos) porque son puntos compartidos entre las diferentes párabolas creadas. 

Finalmente, normalizamos el resultado multiplicando por $\frac {h}{3}$, pues es un ajuste necesario al integrar la función en un intervalo de tamaño **$2h$** y normalizarlo para que funcione en cada intervalo creado. Así pues, tenemos un método de **integración** que surge de integrar funciones por tramos parabólicos e interpretar matemáticamente el peso o incidencia de cada punto de evaluación.

Ahora bien, explicado el método de integración, podemos ir al meollo del asunto: **¿es computable esta forma de integrar?**, **¿es óptimo?**, **¿paralelizable?**, probémoslo, escribamos un programa en C que realice estos cálculos aplicados a la integral previamente planteada y respondamos a estas cuestiones:

In [14]:
//%ldflags: -lm
#include <stdio.h>
#include <math.h>
#include <omp.h>
#define real_value 1.4936490014
// Line used to define pi.

double function(double x) {
    return exp(-(x*x));
}

double calculate_integral_using_simpson_one_third(double inferior_limit, double superior_limit, double pairs) {
    if (fmod(pairs, 2) != 0) {
        printf("La cantidad de Intervalos ingresada NO es válida.");
    } else {
        double even_summatorie = 0.0;
        double odd_summatorie = 0.0;
        double h = (superior_limit - inferior_limit) / pairs;
        double x_0 = function(inferior_limit);
        double x_n = function(superior_limit);
    
        int i = 1;
        for (i = 1; i < (pairs); i++) {
            if ((i % 2) == 0) {
                double x_i = inferior_limit + (i * h);
                even_summatorie += function(x_i);
            } else {
                double x_i = inferior_limit + (i * h);
                odd_summatorie += function(x_i);
            }
        }

        double normalizer = (h / 3);
        double calculated_area = x_0 + (4 * odd_summatorie) + (2 * even_summatorie) + x_n;
        double final_area = normalizer * calculated_area; 
        return final_area;
    }
}

int main() {
    double inferior_limit = -1;
    double superior_limit = 1;
    double pairs[6] = {20.0, 60.0, 500.0, 1000.0, 100000.0, 1000000.0};

    FILE *file = fopen("data2.dat", "w");
    fprintf(file, "pares tiempo_ms resultado error\n");

    int i = 1;
    for (i = 0; i < 6; i++) {
        double initial_time = omp_get_wtime();
        double area = calculate_integral_using_simpson_one_third(inferior_limit, superior_limit, pairs[i]);
        double error = fabs(real_value - area);
        double final_time = omp_get_wtime();
        double time_passed = (final_time - initial_time) * 1000.0;

        printf("Si hallamos el área de la función cos(x²) en el intervalo [%.1f, %.1f] logramos: %.10f\n", inferior_limit, superior_limit, area);
        printf("Con %.1f iteraciones, el tiempo de ejecución fue de %f ms \n", pairs[i], time_passed);
        printf("\n");

        fprintf(file, "%.0f %.6f %.10f %.10f\n", pairs[i], time_passed, area, error);
    }

    fclose(file);
    return 0;
}

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936498965
Con 20.0 iteraciones, el tiempo de ejecución fue de 0.085365 ms 

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482858
Con 60.0 iteraciones, el tiempo de ejecución fue de 0.087902 ms 

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482656
Con 500.0 iteraciones, el tiempo de ejecución fue de 0.034992 ms 

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482656
Con 1000.0 iteraciones, el tiempo de ejecución fue de 0.042755 ms 

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482656
Con 100000.0 iteraciones, el tiempo de ejecución fue de 5.024095 ms 

Si hallamos el área de la función cos(x²) en el intervalo [-1.0, 1.0] logramos: 1.4936482656
Con 1000000.0 iteraciones, el tiempo de ejecución fue de 61.340097 ms 



Primero, definimos, nuevamente las cabeceras de compilación `<stdio.h>`, `<math.h>` y `<omp.h>`, junto con una constante de compilación `real_value`, una variable que contiene el valor real de la integral (importante para realizar comparaciones y medir el error absoluto de cada iteración). Luego definimos una función `double function(double x) {}` en la que definimos la función a integrar, pero en este caso simplemente se recibe un valor $x$, y se calcula la imagen $y$ de la función en ese punto Posterior a eso, se implementó una función: 

`double calculate_integral_using_simpson_one_third(double inferior_limit, double superior_limit, double pairs) {}`

En la que se hace el cálculo integral de un área específica. Inicialmente, en la función recibimos 3 parámetros:

1. **Límite Inferior de Integración**.
2. **Límite Superior de Integración**.
3. **La cantidad de Pares en la que queremos dividir el intervalo.**

Posterior a esto, tenemos una validación simple, los pares deben ser un **número par** de divisiones, de no ser así, no se ejecuta nada. Si se cumple con esta restricción, definimos una variable para almacenar la sumatoria, tanto de nodos **pares**, como de nodos **impares** (`double even_summatorie` y `double odd_summatorie`, respectivamente), una variable para definir la distancia de cada par (`h`, en este caso), y luego, evaluar la función en los límites de integración para utilizarlos en el cálculo final (recordemos que son valores con peso uno, que deben calcularse).

Finalmente, definimos un índice auxiliar `int i` cuya función va a ser servir para un ciclo `for` en el que se realizarán las sumatorias de cada uno de los nodos, tanto **pares** como **impares**, iniciando desde 1. Si el nodo es par, se calcula su distancia h utilizando el límite, y esa distancia es utilizada para meterla dentro de la función original ($e^{-x²}$), el resultado logrado a partir de eso se suma en la variable `even_summatorie`; si el nodo es **impar** se realiza el mismo procedimiento, solo que se añade el valor conseguido a la variable `odd_summatorie`. Al final, se halla $\frac{h}{3}$, y se multiplica por la suma de los **dos términos de los extremos**, 4 veces la sumatoria de **impares** y 2 veces la sumatoria de **pares**. De esta forma, logramos y retornamos el valor calculado en la función.

Ahora, **¿esto funciona?**, veamos y analicemos los resultados; se inicializó, en la función `main() {}`, una serie de variables, primero, para los límites superior e inferior de la integral buscada ($[-1, 1]$)

$$\int_{-1}^{1} e^{-x²} \ dx$$

Luego, utilizamos un **arreglo unidimensional** en el que almacenaremos todos los pares con los que se probará el método (es un arreglo que va desde 20 divisiones, hacia un total de un millón), así pues, con un ciclo simple `for`, automatizamos la respuesta de el algoritmo por cada una de las divisiones en el arreglo. Al final, vemos que luego de solo **20 divisiones**, logramos un resultado de $1.4936498965 [u²]$, y que luego de un millón aproximaciones, el resultado es de $1.4936482656 [u²]$. 

Aproximaciones excelentes, incluso desde números muy bajos en divisiones, efectivamente, nuestro algoritmo funciona; es, nuevamente, un algoritmo simple, pues no se realizan recorridos **costosos en memoria**, es decir, se realizan operaciones básicas, y 2 sumatorias cuya complejidad algorítmica es, de nuevo, lineal; al igual que la **Sumatoria de Riemann**, tenemos un método de **Integración Numérica** computable y poco complejo (o demandante) para un procesador. Ahora, notemos otra cosa, y es que la diferencia con respecto a el otro método, es abrupta: ya desde las 20 iteraciones teníamos una aproximación excelente, y además de eso, la convergencia dejo de ser notable desde las **500 iteraciones**, es decir, el algoritmo nos dió la respuesta que buscamos en un número de iteraciones mucho menor al utilizado en Riemann; igualmente, notar que aunque en **20 divisiones** logramos un resultado muy exacto, si aumentamos hasta **500**, el resultado es el esperado: al aumentar las iteraciones (hasta cierto punto), nos permite llegar a una mejor aproximación.

Esto ocurre porque por **definición** del método, a medida de que $h$ se va haciendo más pequeño, el método mejora en su aproximación, porque el error del método se relaciona **directamente** con la cuarta potencia de $h$ ($O(h⁴)$); al hacer más y más divisiones, la longitud de cada par de sub-intervalos es cada vez menor, y al ser cada vez menor, **es mucho más fácil y preciso el ajuste de la parábola a la curva**. 

En síntesis, tenemos un método que funciona en sus aproximaciones, tiene tiempos **ideales** de ejecución para **iteraciones altas**, y es óptimo tanto en **espacio de memoria**, como en **complejidad algoritmica**. Sin embargo, comparemos su rendimiento con respecto a los resultados obtenidos utilizando **Riemann**: **¿es este método mejor en sus aproximaciones?**, **¿Sí converge más rápido que el anterior?**, **¿Cuál es mejor en sus tiempos?**; revisemos las siguientes gráficas de rendimiento para notar comportamientos y confirmar hipótesis: 

<div style="text_align = center;">
  <img src="tiempo_vs_pares.png" width="400"/>
  <img src="convergencia_vs_pares.png" width="400"/>
</div>

Al ver la comparación entre la **cantidad de pares**, y **tiempo de ejecución**, confirmamos que el comportamiento del método, pose una complejidad lineal $O(n)$ (despreciando las iteraciones más bajas), pues cada una de las iteraciones, tiene un tiempo **proporcional** al anterior. Además, la segunda visualización de convergencia, nos confirma, de forma gráfica, lo dicho anteriormente, la convergencia aumenta al aumentar la cantidad de pares, con un comportamiento de estabilización acelerado. Ahora, comparemos el **comportamiento** de los algoritmos anteriormente planteados, veamos las siguientes gráficas y tabla:

<div style="text_align = center;">
  <img src="comparacion_tiempo.png" width="400"/>
  <img src="comparacion_convergencia.png" width="400"/>
</div>

Aquí vemos, fascinante y claramente, la diferencia entre los dos algoritmos, si analizamos el tiempo de ejecución de ambos, en iteraciones pequeñas, **Simpson** toma un tiempo considerablemente mayor, luego, con divisiones mucho mayores, **Simpson** sigue tomando un tiempo mayor en su ejecución que **Riemann**, ¿por qué?, Simple, cuando ejecutamos el ciclo de suma utilizando el método de **Simpson 1/3**, realizamos operaciones un poco más demandantes, pues no solo realizamos las sumas iteradas, si no que además de eso calculamos **la imagen** de los **puntos inicial y final** del intervalo en la función, multiplicamos las sumatorias por sus respectivos pesos, y además para hacer las sumatorias de pares e impares, debemos realizar un desplazamiento del valor; como vemos, es una cantidad de operaciones mayor a la de **Riemann**, pues en el primer método numérico simplemente evaluamos, multiplicamos por la longitud de cada división **n**, y sumamos. 

**Simpson 1/3** es un algoritmo simple, sí, pero aún siendo simple, es un poco más demandante que el otro método tratado, por la cantidad mayor de operaciones a realizar. Sin embargo, ¿vale la pena que esto sea así?

Totalmente, y es que, en la otra gráfica (y la más diciente en términos comparativos) logramos ver las diferencias de convergencia entre ambos métodos, y aquí, la eficiencia de **Simpson 1/3** es arrolladora, no es solo el hecho de que su error sea bajo, es el hecho de que, incluso desde las 20 iteraciones, vemos un resultado **mucho más cercano** al valor de la integral, que en las **60 iteraciones** de Riemann; la comparación es clara, a pesar de que **Simpson** pueda llegar a ser un poco más demandante, las recompensas son gigantes, pues la convergencia ya se da, incluso desde un número bajo de iteraciones: **Simpson** converge muchísimo más rapido por iteración que **Riemann**, y deja de tener relevancia su convergencia desde muchas iteraciones antes que **Riemann**.

En el primer método, al seguir iterando, se puede mejorar el error de la aproximación incluso luego de cien mil rectángulos, con **Simpson**, deja de ser necesario iterar desde las **500 divisiones**, es una diferencia en rendimiento brutal y muy clara. Confirmemos lo que vemos con el siguiente cuadro comparativo:

| Iteraciones / Pares | Riemann: Resultado | Riemann: Error | Riemann: Tiempo (ms) | Simpson 1/3: Resultado | Simpson 1/3: Error | Simpson 1/3: Tiempo (ms) |
| ------------------: | -----------------: | -------------: | -------------------: | ---------------------: | -----------------: | -----------------------: |
|                  20 |       1.5222413202 |   0.0285923188 |             0.035567 |           1.4936498965 |       0.0000008951 |                 0.085365 |
|                  60 |       1.5049710625 |   0.0113220611 |             0.004031 |           1.4936482858 |       0.0000007156 |                 0.087902 |
|                 500 |       1.4936463036 |   0.0000026978 |             0.019140 |           1.4936482656 |       0.0000007358 |                 0.034992 |
|                1000 |       1.4936477751 |   0.0000012263 |             0.037508 |           1.4936482656 |       0.0000007358 |                 0.042755 |
|              100000 |       1.4936482656 |   0.0000007358 |             3.919010 |           1.4936482656 |       0.0000007358 |                 5.024095 |
|             1000000 |       1.4936490014 |   0.0000000000 |            56.464780 |           1.4936482656 |       0.0000007358 |                61.340097 |


Así pues, concluímos algo muy claro, y es que **Simpson 1/3** logra errores del orden de **$10^{-7}$** desde unas pocas decenas de **divisiones pares**, para llegar a ese nivel de convergencia, **Riemann** necesita más de mil. Así pues, logramos un algoritmo mucho más óptimo en términos de tiempo y de consecución de objetivos que el anterior, utilizando operaciones simples y recorridos también muy simples de ejecutar computacionalmente.

Sin embargo, a algo que ya demostramos que **mejora las aproximaciones de forma abrupta**, ¿podemos hacerlo incluso más óptimo?, ¿podemos mejorar incluso más la forma en la que lo implementamos? La respuesta es que sí, el algoritmo puede ser incluso mejor, a pesar de que **de por sí**, su uso ya nos mejora muchísimo la capacidad de cálculo.

Pero entonces, **¿qué podemos mejorar?** Pensemos en esto: el método debe realizar sumatorias para evaluar la función en cada nodo, sumatorias las cuales, a **divisiones pares** enormes, se vuelven extremadamente demandantes, pues no es simplemente una suma, se evalúa un **punto en la función** y se añade a la **sumatoria**, agregale a eso las operaciones finales de aproximación, las operaciones no son tan demandantes en términos de procesador, pero, **¿y si las volvemos incluso menos demandantes?**

Ahí es donde entra un factor que nos puede hacer la vida mucho más facil: **la paralelización**, **¿y si logramos, que en vez de que el procesador haga una única sumatoria, se pueda dividir el trabajo, y realizarlo simultáneamente en diferentes hilos de ejecución?**, es una idea excelente, pues hacemos que las acumulaciones se dividan, se calculen de forma paralela e independiente, y al finalizar, acumulamos cada parte. Ahora, ¿cómo podemos lograr esto?, para esas funciones existe **OpenMP**, con sus directivas de compilación, podemos paralelizar el cálculo, y lograr una mayor velocidad en las sumatorias. Realicemos el ejercicio implementando un programa en C que permita realizar integraciones numéricas utilizando **Simpson 1/3**, pero paralelizando esas operaciones para una mayor velocidad: